### ISSUE TITLE
Implement true preemption for managed code

### ISSUE THREAD

**Issue by sporksmith** [shadow/shadow#2066]
A thread that never makes any syscalls will stop a Shadow simulation from making progress.

We haven't encountered this in practice, but have encountered other "busy loops" that wait for another thread to do something. Our new feature for modeling latency of non-blocking syscalls helps us escape most such loops, and arguably it'd be pretty poor code that didn't have some kind of timeout, which would imply a system call to check the time.

Still, it's not too far-fetched to imagine, e.g. spawning a thread and then busy-waiting for the new thread to modify an atomic in memory. Such code probably *ought* to have a timeout or at least a `sched_yield`, but undoubtedly code exists with neither.

This is currently low priority, but let's track any real-world instances of this problem that we do find here.

**Comment by sporksmith** [comment]
I think it wouldn't be terribly difficult to implement preemption using [setitimer](https://man7.org/linux/man-pages/man2/setitimer.2.html) to arrange  for an interrupt after some amount of CPU time, and having out interrupt handler move time forward and potentially reschedule the thread. This would definitely break determinism, though.

**Comment by sporksmith** [comment]
> We haven't encountered this in practice, 

We have, now.

* [spin crate](https://github.com/shadow/shadow/issues/1792#issuecomment-1442987642)
* [mysql server](https://github.com/shadow/shadow/issues/1792#issuecomment-2223190467)

**Comment by sporksmith** [comment]
> I think it wouldn't be terribly difficult to implement preemption using [setitimer](https://man7.org/linux/man-pages/man2/setitimer.2.html) 

A tricky bit with `setitimer` is that it sends process-targeted signals, so we'd need to be careful to block the signal in all non-running threads to ensure it gets delivered to the running (busy-looping) thread.

It looks like `timer_create` supports sending thread-targeted signals, *and* a clock that measures cpu-time of the current thread (`CLOCK_THREAD_CPUTIME_ID`). Using this API we won't have to worry about blocking and unblocking the signal. We may still want to reset the timer when rescheduling the thread, but could probably get away with not doing so. e.g. we could just track the last simulated time that the timer went off, and only actually do a preemption if it fires again and simulated time hasn't advanced.

**Comment by sporksmith** [comment]
My current prototype is using `setitimer` to avoid the potential resource limits and "here be dragons" warnings  for `timer_create + SIGEV_THREAD_ID`.

@stevenengler points out that we could avoid the resource limit issue with `timer_create` by completely destroying the timer when switching threads, though we might want to investigate the performance tradeoff of doing that vs pausing/resetting without destroying.

**Comment by sporksmith** [comment]
In case we revisit this - I just stumbled across  `RLIMIT_CPU`, which could also be used for this, though I'm not sure  there's any advantage over the itimer-based approach. Maybe if it can send thread-targeted signals instead of process-targeted; the docs seem to suggest process-oriented, but a lot of linux docs use "process" and "thread" interchangeably.

`setrlimit(2)`:

> RLIMIT_CPU
> This is a limit, in seconds, on the amount of CPU time that the  process  can  consume.   When  the  process reaches the soft limit, it is sent a SIGXCPU signal.  The default action for this signal is to terminate the process.  However, the signal can be caught, and the handler can return control to the main program.  If the process  continues  to  consume  CPU  time,  it will be sent SIGXCPU once per second until the hard limit is reached, at which time it is sent SIGKILL.  (This latter point describes  Linux  behavior.   Implementations vary in how they treat processes which continue to consume CPU time after reaching the soft limit.  Portable applications that need to catch this signal should perform an orderly  termination  upon  first  receipt  of SIGXCPU.)
